nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,344 @@
+  slowdown_pub_ = node->create_publisher<visualization_msgs::msg::Marker>("slowdown", 1);
+
+  RCLCPP_INFO(logger_, "Configured Graceful Motion Controller: %s", plugin_name_.c_str());
+}
+
+void GracefulMotionController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_.reset();
+  local_plan_pub_.reset();
+  motion_target_pub_.reset();
+  slowdown_pub_.reset();
+}
+
+void GracefulMotionController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_->on_activate();
+  local_plan_pub_->on_activate();
+  motion_target_pub_->on_activate();
+  slowdown_pub_->on_activate();
+}
+
+void GracefulMotionController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_->on_deactivate();
+  local_plan_pub_->on_deactivate();
+  motion_target_pub_->on_deactivate();
+  slowdown_pub_->on_deactivate();
+}
+
+geometry_msgs::msg::TwistStamped GracefulMotionController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & /*velocity*/,
+  nav2_core::GoalChecker * goal_checker)
+{
+  std::lock_guard<std::mutex> param_lock(param_handler_->getMutex());
+
+  // Update for the current goal checker's state
+  geometry_msgs::msg::Pose pose_tolerance;
+  geometry_msgs::msg::Twist velocity_tolerance;
+  if (!goal_checker->getTolerances(pose_tolerance, velocity_tolerance)) {
+    RCLCPP_WARN(logger_, "Unable to retrieve goal checker's tolerances!");
+  } else {
+    goal_dist_tolerance_ = pose_tolerance.position.x;
+  }
+
+  // Transform path to robot base frame and publish it
+  auto transformed_plan = path_handler_->transformGlobalPlan(
+    pose, params_->max_robot_pose_search_dist);
+  transformed_plan_pub_->publish(transformed_plan);
+
+  // Get the particular point on the path at the motion target distance and publish it
+  auto motion_target = getMotionTarget(params_->motion_target_dist, transformed_plan);
+  auto motion_target_point = nav2_graceful_motion_controller::createMotionTargetMsg(motion_target);
+  motion_target_pub_->publish(motion_target_point);
+
+  // Publish marker for slowdown radius around motion target for debugging / visualization
+  auto slowdown_marker = nav2_graceful_motion_controller::createSlowdownMarker(
+    motion_target,
+    params_->slowdown_radius);
+  slowdown_pub_->publish(slowdown_marker);
+
+  // Compute distance to goal
+  geometry_msgs::msg::PoseStamped goal_pose = transformed_plan.poses.back();
+  double dist_to_goal = std::hypot(goal_pose.pose.position.x, goal_pose.pose.position.y);
```

Comment:
Might want to use the path's integrated distance, there's a util for that in nav2_utils. That way its not just the L2 distance, but the integrated path distances to account for path curvatures